src/interface/variable_pack.hpp

```diff
@@ -49,6 +49,9 @@ using SwarmVarList = std::forward_list<std::shared_ptr<ParticleVariable<T>>>;
 // the pairs represent interval (inclusive) of those indices
 using IndexPair = std::pair<int, int>;
 
+// Used for storing the shapes of variable fields
+using Shape = std::vector<int>;
```

Comment:
If we wanted to be able to put things on device, this could be a class that owns a `ParArray1D<int>` and the `GetFlatIdx` method. And you could get that object for a given key.

Reply:
I have no strong preference, although I think the flat indices should probably be pre-computed on the host. I guess this could be done during construction in your suggested `FlatIdx` class, although I guess the that would require a `ParArrayND` member of indices rather than just a `ParArray1D` containing the shape .

Reply:
If you like them being precomputed on the host, then I think the way you've written it is perfectly good.

Reply:
hm, looking at one of the use cases I have a little bit more, there might be some advantage to calculating the FlatIdx on the device since the indexing array could be reasonably large. I am going to revert back to WIP and implement your suggestion since it is more flexible.

Reply:
Sounds good.

Reply:
Ping me when you're ready for re-review

Reply:
It is ready to be looked at again. I ran it through tests on my machine and everything works as expected, but I have only tested purely on CPU.